service/kvmclusterconfig/v1/resource/encryptionkey/create.go

```diff
@@ -11,7 +11,21 @@ import (
 // Patch provided by NewUpdatePatch or NewDeletePatch. It creates k8s secret
 // for encryption key if needed.
 func (r *Resource) ApplyCreateChange(ctx context.Context, obj, createChange interface{}) error {
-	return nil
+	secret, err := toSecret(createChange)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if secret == nil {
+		return microerror.Maskf(invalidValueError, "createChange (*v1.Secret) must not be nil")
+	}
```

Comment:
```
	var secretToCreate *v1.Secret
	if currentSecret == nil {
		secretToCreate = desiredSecret
	}

	r.logger.LogCtx(ctx, "debug", "found out if the secret has to be created")

	return secretToCreate, nil
```

So if `currentSecret  != nil`, then `secretToCreate` (aka. createChange) is nil.

Am I missing something?

Reply:
Oh, right. So this is now the case that I wanted to avoid in the first place with the condition in `NewUpdatePatch`. I'm quite strongly against this pattern as for me at least this is illogical considering the mechanisms in our framework and I feel this is also a bit error prone as seen here. From what I can understand here https://github.com/giantswarm/operatorkit/blob/master/framework/framework.go#L485 if there were conditionals in `NewUpdatePatch`, we would run through less unnecessary code (therefore less possibilities for bugs as well).

Anyway, I'll change the logic here so that it's in line with other repositories.

Reply:
I am not sure what Pawel wants to say but I can say that the operator is executing the creation magic continuously and we cannot expect the create state to be always given, because that would imply once created the resource we are managing here would be deleted after we created it. So we should just ensure the creation and be fine in case there is nothing to create because this means we did our job and be done.

Reply:
I want to say, secret can be nil here and this is a valid case.

> I'm quite strongly against this pattern as for me at least this is illogical considering the mechanisms in our framework and I feel this is also a bit error prone as seen here.

I understand that. This is also how the Patch struct was designed. Problem here is logging which @xh3b4sd finds beneficial. Maybe they can be moved to the NewPatch* functions.

Reply:
I am not with you that this is error prone because AFAIK we never had problems with this. The framework is open and lets you decide what to do without a lot of magic under the hood. Adding conditionals to code the user never sees will lead to more problems than you try to solve. We can for sure always talk about this but I just do not see how we can improve this here as of now. The framework should just be as transparent as possible and I think it is a good thing for the user to decide what to do when the zero value of its dispatched state is received.